improve escaped-quote handling in sys.process (CommandLineParser)

SethTisue opened the pull request:
fixes scala/bug#12611

SethTisue committed:
improve escaped-quote handling in sys.process and tools.cmd

fixes scala/bug#12611

som-snytt commented on src/compiler/scala/tools/cmd/CommandLineParser.scala:
```diff
@@ -14,6 +14,11 @@ package scala.tools.cmd
+// and should be maintained in parallel. (Ideally we'd eliminate
+// the duplication, which arose because we expected sys.process to
+// become a separate module.)
+
```
`object CommandLineParser extends scala.sys.process.Parser`

som-snytt commented on src/library/scala/sys/process/Parser.scala:
```diff
@@ -14,6 +14,11 @@ package scala.sys.process
+
 /** A simple enough command line parser.
  */
 private[scala] object Parser {
```
```
-private[scala] object Parser {
+private[scala] class Parser {

-      def terminal = in.charAt(pos) match {
+      def terminal: Boolean = cur match {

-    @tailrec
-    def skipToDelim(): Boolean =
-      cur match {
+    // Skip to a word boundary, where words can be quoted and quotes can be escaped
+    def skipToDelim(): Boolean = {
+      var escaped = false
+      def advance(): Boolean = cur match {
+        case _ if escaped         => escaped = false ; bump() ; advance()
+        case '\\'                 => escaped = true ; bump() ; advance()
         case q @ (DQ | SQ)        => { qpos += pos; bump(); skipToQuote(q) } && { qpos += pos; bump(); skipToDelim() }
         case -1                   => true
         case c if isWhitespace(c) => true
         case _                    => bump(); skipToDelim()
       }
+      advance()
+    }

+object Parser extends Parser
```
I could push to branch if the tweaks look good.

som-snytt commented on src/library/scala/sys/process/Parser.scala:
```diff
@@ -49,9 +54,12 @@ private[scala] object Parser {
       while (!done && !terminal) pos += 1
       !done
     }
+    var escapedCur = false
```
My suggestion makes it more local.

som-snytt reviewed (approved):
So much for "at least I'll never have to look at that code ever again."

I could push tweaks to branch or follow up when I forward port it.

som-snytt committed:
CommandLineParser is process.Parser

Also backport tweaks from dotty.

CommandLineParser is no more

DirectTests, which are not under package scala,
inherit tokenize from DirectTest.

som-snytt commented:
Without waiting for the slow green check, I will also push Seth's suggestion to bite the bullet and remove the `cmd` package. There are just a handful of usages.

som-snytt commented:
Squashed my commits, including removal of `tools.cmd`, because of funky mima -- it complained about missing ctors for `ParseException`, when `class Parser {...}; object Parser extends Parser`. That form was unnecessary after removing `CommandLineParser extends Parser`.

Fixed tests which "leveraged" `CommandLineParser`.

som-snytt commented:
Requires Seth's review of following up his suggestions on the ticket and thereby creating more work for himself.
